##### Copyright 2019 The TensorFlow Authors.

In [1]:
#@title Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

In [2]:
#@title MIT License
#
# Copyright (c) 2017 François Chollet
#
# Permission is hereby granted, free of charge, to any person obtaining a
# copy of this software and associated documentation files (the "Software"),
# to deal in the Software without restriction, including without limitation
# the rights to use, copy, modify, merge, publish, distribute, sublicense,
# and/or sell copies of the Software, and to permit persons to whom the
# Software is furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
# THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
# DEALINGS IN THE SOFTWARE.

# Text classification with TensorFlow Hub: Movie reviews

<table class="tfo-notebook-buttons" align="left">
  <td>
    <a target="_blank" href="https://www.tensorflow.org/tutorials/keras/text_classification_with_hub"><img src="https://www.tensorflow.org/images/tf_logo_32px.png" />View on TensorFlow.org</a>
  </td>
  <td>
    <a target="_blank" href="https://colab.research.google.com/github/tensorflow/docs/blob/master/site/en/tutorials/keras/text_classification_with_hub.ipynb"><img src="https://www.tensorflow.org/images/colab_logo_32px.png" />Run in Google Colab</a>
  </td>
  <td>
    <a target="_blank" href="https://github.com/tensorflow/docs/blob/master/site/en/tutorials/keras/text_classification_with_hub.ipynb"><img src="https://www.tensorflow.org/images/GitHub-Mark-32px.png" />View on GitHub</a>
  </td>
  <td>
    <a href="https://storage.googleapis.com/tensorflow_docs/docs/site/en/tutorials/keras/text_classification_with_hub.ipynb"><img src="https://www.tensorflow.org/images/download_logo_32px.png" />Download notebook</a>
  </td>
  <td>
    <a href="https://tfhub.dev/s?module-type=text-embedding"><img src="https://www.tensorflow.org/images/hub_logo_32px.png" />See TF Hub models</a>
  </td>
</table>

This notebook classifies movie reviews as *positive* or *negative* using the text of the review. This is an example of *binary*—or two-class—classification, an important and widely applicable kind of machine learning problem.

The tutorial demonstrates the basic application of transfer learning with [TensorFlow Hub](https://tfhub.dev) and Keras.

It uses the [IMDB dataset](https://www.tensorflow.org/api_docs/python/tf/keras/datasets/imdb) that contains the text of 50,000 movie reviews from the [Internet Movie Database](https://www.imdb.com/). These are split into 25,000 reviews for training and 25,000 reviews for testing. The training and testing sets are *balanced*, meaning they contain an equal number of positive and negative reviews. 

This notebook uses [`tf.keras`](https://www.tensorflow.org/guide/keras), a high-level API to build and train models in TensorFlow, and [`tensorflow_hub`](https://www.tensorflow.org/hub), a library for loading trained models from [TFHub](https://tfhub.dev) in a single line of code. For a more advanced text classification tutorial using `tf.keras`, see the [MLCC Text Classification Guide](https://developers.google.com/machine-learning/guides/text-classification/).

In [3]:
!pip install tensorflow-hub
!pip install tensorflow-datasets

In [4]:
import os
import numpy as np

import tensorflow as tf
import tensorflow_hub as hub
import tensorflow_datasets as tfds

print("Version: ", tf.__version__)
print("Eager mode: ", tf.executing_eagerly())
print("Hub version: ", hub.__version__)
print("GPU is", "available" if tf.config.list_physical_devices("GPU") else "NOT AVAILABLE")

Version:  2.8.0-rc1
Eager mode:  True
Hub version:  0.12.0
GPU is available


## Download the IMDB dataset

The IMDB dataset is available on [imdb reviews](https://www.tensorflow.org/datasets/catalog/imdb_reviews) or on [TensorFlow datasets](https://www.tensorflow.org/datasets). The following code downloads the IMDB dataset to your machine (or the colab runtime):

In [5]:
# Split the training set into 60% and 40% to end up with 15,000 examples
# for training, 10,000 examples for validation and 25,000 examples for testing.
train_data, validation_data, test_data = tfds.load(
    name="imdb_reviews", 
    split=('train[:60%]', 'train[60%:]', 'test'),
    as_supervised=True)

## Explore the data 

Let's take a moment to understand the format of the data. Each example is a sentence representing the movie review and a corresponding label. The sentence is not preprocessed in any way. The label is an integer value of either 0 or 1, where 0 is a negative review, and 1 is a positive review.

Let's print first 10 examples.

In [6]:
train_examples_batch, train_labels_batch = next(iter(train_data.batch(10)))
train_examples_batch

<tf.Tensor: shape=(10,), dtype=string, numpy=
array([b"This was an absolutely terrible movie. Don't be lured in by Christopher Walken or Michael Ironside. Both are great actors, but this must simply be their worst role in history. Even their great acting could not redeem this movie's ridiculous storyline. This movie is an early nineties US propaganda piece. The most pathetic scenes were those when the Columbian rebels were making their cases for revolutions. Maria Conchita Alonso appeared phony, and her pseudo-love affair with Walken was nothing but a pathetic emotional plug in a movie that was devoid of any real meaning. I am disappointed that there are movies like this, ruining actor's like Christopher Walken's good name. I could barely sit through it.",
       b'I have been known to fall asleep during films, but this is usually due to a combination of things including, really tired, being warm and comfortable on the sette and having just eaten a lot. However on this occasion I fell 

Let's also print the first 10 labels.

In [7]:
train_labels_batch

<tf.Tensor: shape=(10,), dtype=int64, numpy=array([0, 0, 0, 1, 1, 1, 0, 0, 0, 0])>

## Build the model

The neural network is created by stacking layers—this requires three main architectural decisions:

* How to represent the text?
* How many layers to use in the model?
* How many *hidden units* to use for each layer?

In this example, the input data consists of sentences. The labels to predict are either 0 or 1.

One way to represent the text is to convert sentences into embeddings vectors. Use a pre-trained text embedding as the first layer, which will have three advantages:

*   You don't have to worry about text preprocessing,
*   Benefit from transfer learning,
*   the embedding has a fixed size, so it's simpler to process.

For this example you use a **pre-trained text embedding model** from [TensorFlow Hub](https://tfhub.dev) called [google/nnlm-en-dim50/2](https://tfhub.dev/google/nnlm-en-dim50/2).

There are many other pre-trained text embeddings from TFHub that can be used in this tutorial:

* [google/nnlm-en-dim128/2](https://tfhub.dev/google/nnlm-en-dim128/2) - trained with the same NNLM architecture on the same data as [google/nnlm-en-dim50/2](https://tfhub.dev/google/nnlm-en-dim50/2), but with a larger embedding dimension. Larger dimensional embeddings can improve on your task but it may take longer to train your model.
* [google/nnlm-en-dim128-with-normalization/2](https://tfhub.dev/google/nnlm-en-dim128-with-normalization/2) - the same as [google/nnlm-en-dim128/2](https://tfhub.dev/google/nnlm-en-dim128/2), but with additional text normalization such as removing punctuation. This can help if the text in your task contains additional characters or punctuation.
* [google/universal-sentence-encoder/4](https://tfhub.dev/google/universal-sentence-encoder/4) - a much larger model yielding 512 dimensional embeddings trained with a deep averaging network (DAN) encoder.

And many more! Find more [text embedding models](https://tfhub.dev/s?module-type=text-embedding) on TFHub.

Let's first create a Keras layer that uses a TensorFlow Hub model to embed the sentences, and try it out on a couple of input examples. Note that no matter the length of the input text, the output shape of the embeddings is: `(num_examples, embedding_dimension)`.

In [8]:
embedding = "https://tfhub.dev/google/nnlm-en-dim50/2"
hub_layer = hub.KerasLayer(embedding, input_shape=[], 
                           dtype=tf.string, trainable=True)
hub_layer(train_examples_batch[:3])

<tf.Tensor: shape=(3, 50), dtype=float32, numpy=
array([[ 0.5423194 , -0.01190171,  0.06337537,  0.0686297 , -0.16776839,
        -0.10581177,  0.168653  , -0.04998823, -0.31148052,  0.07910344,
         0.15442258,  0.01488661,  0.03930155,  0.19772716, -0.12215477,
        -0.04120982, -0.27041087, -0.21922147,  0.26517656, -0.80739075,
         0.25833526, -0.31004202,  0.2868321 ,  0.19433866, -0.29036498,
         0.0386285 , -0.78444123, -0.04793238,  0.41102988, -0.36388886,
        -0.58034706,  0.30269453,  0.36308962, -0.15227163, -0.4439151 ,
         0.19462997,  0.19528405,  0.05666233,  0.2890704 , -0.28468323,
        -0.00531206,  0.0571938 , -0.3201319 , -0.04418665, -0.08550781,
        -0.55847436, -0.2333639 , -0.20782956, -0.03543065, -0.17533456],
       [ 0.56338924, -0.12339553, -0.10862677,  0.7753425 , -0.07667087,
        -0.15752274,  0.01872334, -0.08169781, -0.3521876 ,  0.46373403,
        -0.08492758,  0.07166861, -0.00670818,  0.12686071, -0.19326551,
 

Let's now build the full model:

In [9]:
model = tf.keras.Sequential()
model.add(hub_layer)
model.add(tf.keras.layers.Dense(16, activation='relu'))
model.add(tf.keras.layers.Dense(1))

model.summary()

Model: "sequential"


_________________________________________________________________


 Layer (type)                Output Shape              Param #   


 keras_layer (KerasLayer)    (None, 50)                48190600  


 dense (Dense)               (None, 16)                816       


 dense_1 (Dense)             (None, 1)                 17        


Total params: 48,191,433


Trainable params: 48,191,433


Non-trainable params: 0


_________________________________________________________________


The layers are stacked sequentially to build the classifier:

1. The first layer is a TensorFlow Hub layer. This layer uses a pre-trained Saved Model to map a sentence into its embedding vector. The pre-trained text embedding model that you are using ([google/nnlm-en-dim50/2](https://tfhub.dev/google/nnlm-en-dim50/2)) splits the sentence into tokens, embeds each token and then combines the embedding. The resulting dimensions are: `(num_examples, embedding_dimension)`. For this NNLM model, the `embedding_dimension` is 50.
2. This fixed-length output vector is piped through a fully-connected (`Dense`) layer with 16 hidden units.
3. The last layer is densely connected with a single output node.

Let's compile the model.

### Loss function and optimizer

A model needs a loss function and an optimizer for training. Since this is a binary classification problem and the model outputs logits (a single-unit layer with a linear activation), you'll use the `binary_crossentropy` loss function.

This isn't the only choice for a loss function, you could, for instance, choose `mean_squared_error`. But, generally, `binary_crossentropy` is better for dealing with probabilities—it measures the "distance" between probability distributions, or in our case, between the ground-truth distribution and the predictions.

Later, when you are exploring regression problems (say, to predict the price of a house), you'll see how to use another loss function called mean squared error.

Now, configure the model to use an optimizer and a loss function:

In [10]:
model.compile(optimizer='adam',
              loss=tf.keras.losses.BinaryCrossentropy(from_logits=True),
              metrics=['accuracy'])

## Train the model

Train the model for 10 epochs in mini-batches of 512 samples. This is 10 iterations over all samples in the `x_train` and `y_train` tensors. While training, monitor the model's loss and accuracy on the 10,000 samples from the validation set:

In [11]:
history = model.fit(train_data.shuffle(10000).batch(512),
                    epochs=10,
                    validation_data=validation_data.batch(512),
                    verbose=1)

Epoch 1/10


 1/30 [>.............................] - ETA: 50s - loss: 0.7396 - accuracy: 0.4473

 3/30 [==>...........................] - ETA: 0s - loss: 0.7317 - accuracy: 0.4616 

 5/30 [====>.........................] - ETA: 0s - loss: 0.7235 - accuracy: 0.4711

 7/30 [======>.......................] - ETA: 0s - loss: 0.7153 - accuracy: 0.4743

 9/30 [========>.....................] - ETA: 0s - loss: 0.7104 - accuracy: 0.4818

12/30 [===========>..................] - ETA: 0s - loss: 0.7036 - accuracy: 0.4884

15/30 [==============>...............] - ETA: 0s - loss: 0.6969 - accuracy: 0.4919

18/30 [=================>............] - ETA: 0s - loss: 0.6906 - accuracy: 0.4977

21/30 [====================>.........] - ETA: 0s - loss: 0.6850 - accuracy: 0.5079

24/30 [=======================>......] - ETA: 0s - loss: 0.6797 - accuracy: 0.5157

27/30 [==========================>...] - ETA: 0s - loss: 0.6747 - accuracy: 0.5249

30/30 [==============================] - ETA: 0s - loss: 0.6707 - accuracy: 0.5317

30/30 [==============================] - 3s 59ms/step - loss: 0.6707 - accuracy: 0.5317 - val_loss: 0.6150 - val_accuracy: 0.5882


Epoch 2/10


 1/30 [>.............................] - ETA: 24s - loss: 0.6079 - accuracy: 0.5938

 3/30 [==>...........................] - ETA: 1s - loss: 0.5893 - accuracy: 0.6178 

 5/30 [====>.........................] - ETA: 0s - loss: 0.5898 - accuracy: 0.6160

 7/30 [======>.......................] - ETA: 0s - loss: 0.5873 - accuracy: 0.6275

 9/30 [========>.....................] - ETA: 0s - loss: 0.5822 - accuracy: 0.6330

12/30 [===========>..................] - ETA: 0s - loss: 0.5759 - accuracy: 0.6466

15/30 [==============>...............] - ETA: 0s - loss: 0.5703 - accuracy: 0.6594

18/30 [=================>............] - ETA: 0s - loss: 0.5652 - accuracy: 0.6687

21/30 [====================>.........] - ETA: 0s - loss: 0.5593 - accuracy: 0.6771

24/30 [=======================>......] - ETA: 0s - loss: 0.5501 - accuracy: 0.6869

27/30 [==========================>...] - ETA: 0s - loss: 0.5437 - accuracy: 0.6951

30/30 [==============================] - ETA: 0s - loss: 0.5382 - accuracy: 0.7012

30/30 [==============================] - 2s 57ms/step - loss: 0.5382 - accuracy: 0.7012 - val_loss: 0.4972 - val_accuracy: 0.7450


Epoch 3/10


 1/30 [>.............................] - ETA: 20s - loss: 0.4463 - accuracy: 0.8125

 3/30 [==>...........................] - ETA: 0s - loss: 0.4350 - accuracy: 0.8132 

 5/30 [====>.........................] - ETA: 0s - loss: 0.4374 - accuracy: 0.8008

 7/30 [======>.......................] - ETA: 0s - loss: 0.4377 - accuracy: 0.7983

 9/30 [========>.....................] - ETA: 0s - loss: 0.4320 - accuracy: 0.8010

12/30 [===========>..................] - ETA: 0s - loss: 0.4258 - accuracy: 0.8078

15/30 [==============>...............] - ETA: 0s - loss: 0.4249 - accuracy: 0.8094

18/30 [=================>............] - ETA: 0s - loss: 0.4177 - accuracy: 0.8128

21/30 [====================>.........] - ETA: 0s - loss: 0.4112 - accuracy: 0.8155

24/30 [=======================>......] - ETA: 0s - loss: 0.4071 - accuracy: 0.8177

27/30 [==========================>...] - ETA: 0s - loss: 0.4019 - accuracy: 0.8209

30/30 [==============================] - ETA: 0s - loss: 0.3976 - accuracy: 0.8246

30/30 [==============================] - 2s 56ms/step - loss: 0.3976 - accuracy: 0.8246 - val_loss: 0.4023 - val_accuracy: 0.8149


Epoch 4/10


 1/30 [>.............................] - ETA: 21s - loss: 0.3405 - accuracy: 0.8672

 3/30 [==>...........................] - ETA: 0s - loss: 0.3280 - accuracy: 0.8750 

 5/30 [====>.........................] - ETA: 0s - loss: 0.3224 - accuracy: 0.8664

 7/30 [======>.......................] - ETA: 0s - loss: 0.3212 - accuracy: 0.8619

 9/30 [========>.....................] - ETA: 0s - loss: 0.3176 - accuracy: 0.8659

12/30 [===========>..................] - ETA: 0s - loss: 0.3086 - accuracy: 0.8747

15/30 [==============>...............] - ETA: 0s - loss: 0.3038 - accuracy: 0.8809

18/30 [=================>............] - ETA: 0s - loss: 0.3000 - accuracy: 0.8826

21/30 [====================>.........] - ETA: 0s - loss: 0.2958 - accuracy: 0.8847

24/30 [=======================>......] - ETA: 0s - loss: 0.2933 - accuracy: 0.8853

27/30 [==========================>...] - ETA: 0s - loss: 0.2905 - accuracy: 0.8869

30/30 [==============================] - ETA: 0s - loss: 0.2879 - accuracy: 0.8894

30/30 [==============================] - 2s 61ms/step - loss: 0.2879 - accuracy: 0.8894 - val_loss: 0.3503 - val_accuracy: 0.8385


Epoch 5/10


 1/30 [>.............................] - ETA: 19s - loss: 0.2195 - accuracy: 0.9219

 3/30 [==>...........................] - ETA: 1s - loss: 0.2383 - accuracy: 0.8965 

 5/30 [====>.........................] - ETA: 0s - loss: 0.2329 - accuracy: 0.9023

 7/30 [======>.......................] - ETA: 0s - loss: 0.2302 - accuracy: 0.9096

 9/30 [========>.....................] - ETA: 0s - loss: 0.2253 - accuracy: 0.9154

12/30 [===========>..................] - ETA: 0s - loss: 0.2251 - accuracy: 0.9183

15/30 [==============>...............] - ETA: 0s - loss: 0.2247 - accuracy: 0.9178

18/30 [=================>............] - ETA: 0s - loss: 0.2216 - accuracy: 0.9192

21/30 [====================>.........] - ETA: 0s - loss: 0.2176 - accuracy: 0.9209

24/30 [=======================>......] - ETA: 0s - loss: 0.2159 - accuracy: 0.9218

27/30 [==========================>...] - ETA: 0s - loss: 0.2127 - accuracy: 0.9239

30/30 [==============================] - ETA: 0s - loss: 0.2120 - accuracy: 0.9243

30/30 [==============================] - 2s 59ms/step - loss: 0.2120 - accuracy: 0.9243 - val_loss: 0.3248 - val_accuracy: 0.8530


Epoch 6/10


 1/30 [>.............................] - ETA: 18s - loss: 0.1580 - accuracy: 0.9609

 3/30 [==>...........................] - ETA: 0s - loss: 0.1631 - accuracy: 0.9473 

 5/30 [====>.........................] - ETA: 0s - loss: 0.1609 - accuracy: 0.9516

 7/30 [======>.......................] - ETA: 0s - loss: 0.1620 - accuracy: 0.9492

 9/30 [========>.....................] - ETA: 0s - loss: 0.1641 - accuracy: 0.9510

12/30 [===========>..................] - ETA: 0s - loss: 0.1621 - accuracy: 0.9495

15/30 [==============>...............] - ETA: 0s - loss: 0.1621 - accuracy: 0.9486

18/30 [=================>............] - ETA: 0s - loss: 0.1582 - accuracy: 0.9506

21/30 [====================>.........] - ETA: 0s - loss: 0.1577 - accuracy: 0.9521

24/30 [=======================>......] - ETA: 0s - loss: 0.1564 - accuracy: 0.9520

27/30 [==========================>...] - ETA: 0s - loss: 0.1559 - accuracy: 0.9515

30/30 [==============================] - ETA: 0s - loss: 0.1560 - accuracy: 0.9513

30/30 [==============================] - 2s 54ms/step - loss: 0.1560 - accuracy: 0.9513 - val_loss: 0.3148 - val_accuracy: 0.8663


Epoch 7/10


 1/30 [>.............................] - ETA: 20s - loss: 0.1272 - accuracy: 0.9629

 3/30 [==>...........................] - ETA: 1s - loss: 0.1275 - accuracy: 0.9635 

 5/30 [====>.........................] - ETA: 0s - loss: 0.1232 - accuracy: 0.9664

 7/30 [======>.......................] - ETA: 0s - loss: 0.1208 - accuracy: 0.9682

 9/30 [========>.....................] - ETA: 0s - loss: 0.1213 - accuracy: 0.9679

11/30 [==========>...................] - ETA: 0s - loss: 0.1219 - accuracy: 0.9668

14/30 [=============>................] - ETA: 0s - loss: 0.1216 - accuracy: 0.9664

17/30 [================>.............] - ETA: 0s - loss: 0.1213 - accuracy: 0.9656

20/30 [===================>..........] - ETA: 0s - loss: 0.1191 - accuracy: 0.9662

23/30 [======================>.......] - ETA: 0s - loss: 0.1170 - accuracy: 0.9673

26/30 [=========================>....] - ETA: 0s - loss: 0.1142 - accuracy: 0.9683

29/30 [============================>.] - ETA: 0s - loss: 0.1147 - accuracy: 0.9679

30/30 [==============================] - 2s 55ms/step - loss: 0.1147 - accuracy: 0.9680 - val_loss: 0.3168 - val_accuracy: 0.8675


Epoch 8/10


 1/30 [>.............................] - ETA: 19s - loss: 0.0848 - accuracy: 0.9922

 3/30 [==>...........................] - ETA: 0s - loss: 0.0974 - accuracy: 0.9837 

 5/30 [====>.........................] - ETA: 0s - loss: 0.0931 - accuracy: 0.9832

 7/30 [======>.......................] - ETA: 0s - loss: 0.0906 - accuracy: 0.9816

 9/30 [========>.....................] - ETA: 0s - loss: 0.0906 - accuracy: 0.9800

12/30 [===========>..................] - ETA: 0s - loss: 0.0894 - accuracy: 0.9792

15/30 [==============>...............] - ETA: 0s - loss: 0.0873 - accuracy: 0.9793

18/30 [=================>............] - ETA: 0s - loss: 0.0862 - accuracy: 0.9799

21/30 [====================>.........] - ETA: 0s - loss: 0.0867 - accuracy: 0.9799

24/30 [=======================>......] - ETA: 0s - loss: 0.0857 - accuracy: 0.9797

27/30 [==========================>...] - ETA: 0s - loss: 0.0851 - accuracy: 0.9789

30/30 [==============================] - ETA: 0s - loss: 0.0847 - accuracy: 0.9791

30/30 [==============================] - 2s 56ms/step - loss: 0.0847 - accuracy: 0.9791 - val_loss: 0.3199 - val_accuracy: 0.8670


Epoch 9/10


 1/30 [>.............................] - ETA: 19s - loss: 0.0700 - accuracy: 0.9883

 3/30 [==>...........................] - ETA: 0s - loss: 0.0715 - accuracy: 0.9824 

 5/30 [====>.........................] - ETA: 0s - loss: 0.0702 - accuracy: 0.9840

 7/30 [======>.......................] - ETA: 0s - loss: 0.0691 - accuracy: 0.9849

10/30 [=========>....................] - ETA: 0s - loss: 0.0674 - accuracy: 0.9865

13/30 [============>.................] - ETA: 0s - loss: 0.0657 - accuracy: 0.9871

16/30 [===============>..............] - ETA: 0s - loss: 0.0643 - accuracy: 0.9878

19/30 [==================>...........] - ETA: 0s - loss: 0.0633 - accuracy: 0.9877

22/30 [=====================>........] - ETA: 0s - loss: 0.0627 - accuracy: 0.9876

25/30 [========================>.....] - ETA: 0s - loss: 0.0622 - accuracy: 0.9874

28/30 [===========================>..] - ETA: 0s - loss: 0.0620 - accuracy: 0.9879

30/30 [==============================] - 2s 54ms/step - loss: 0.0617 - accuracy: 0.9880 - val_loss: 0.3272 - val_accuracy: 0.8649


Epoch 10/10


 1/30 [>.............................] - ETA: 20s - loss: 0.0384 - accuracy: 0.9941

 3/30 [==>...........................] - ETA: 1s - loss: 0.0452 - accuracy: 0.9941 

 5/30 [====>.........................] - ETA: 0s - loss: 0.0460 - accuracy: 0.9941

 7/30 [======>.......................] - ETA: 0s - loss: 0.0457 - accuracy: 0.9941

 9/30 [========>.....................] - ETA: 0s - loss: 0.0445 - accuracy: 0.9948

12/30 [===========>..................] - ETA: 0s - loss: 0.0452 - accuracy: 0.9945

15/30 [==============>...............] - ETA: 0s - loss: 0.0446 - accuracy: 0.9948

18/30 [=================>............] - ETA: 0s - loss: 0.0443 - accuracy: 0.9946

21/30 [====================>.........] - ETA: 0s - loss: 0.0449 - accuracy: 0.9938

24/30 [=======================>......] - ETA: 0s - loss: 0.0448 - accuracy: 0.9938

27/30 [==========================>...] - ETA: 0s - loss: 0.0449 - accuracy: 0.9933

30/30 [==============================] - ETA: 0s - loss: 0.0449 - accuracy: 0.9931

30/30 [==============================] - 2s 57ms/step - loss: 0.0449 - accuracy: 0.9931 - val_loss: 0.3379 - val_accuracy: 0.8651


## Evaluate the model

And let's see how the model performs. Two values will be returned. Loss (a number which represents our error, lower values are better), and accuracy.

In [12]:
results = model.evaluate(test_data.batch(512), verbose=2)

for name, value in zip(model.metrics_names, results):
  print("%s: %.3f" % (name, value))

49/49 - 2s - loss: 0.3599 - accuracy: 0.8523 - 2s/epoch - 34ms/step


loss: 0.360
accuracy: 0.852


This fairly naive approach achieves an accuracy of about 87%. With more advanced approaches, the model should get closer to 95%.

## Further reading

* For a more general way to work with string inputs and for a more detailed analysis of the progress of accuracy and loss during training, see the [Text classification with preprocessed text](./text_classification.ipynb) tutorial.
* Try out more [text-related tutorials](https://www.tensorflow.org/hub/tutorials#text-related-tutorials) using trained models from TFHub.